src/nsrt_learning/strips_learning/gen_to_spec_learner.py

```diff
@@ -269,17 +270,22 @@ def _backchain_one_pass(
                     for nec_pnad in param_opt_to_nec_pnads[option.parent]:
                         pre = self._induce_preconditions_via_intersection(
                             nec_pnad)
+                        self._recompute_datastores_from_segments(
+                            param_opt_to_nec_pnads[option.parent])
                         nec_pnad.op = nec_pnad.op.copy_with(preconditions=pre)
 
                     # After all this, the unification call that failed earlier
                     # (leading us into the current else statement) should work.
-                    best_score_pnad, ground_op = self._find_unification(
-                        necessary_add_effects,
-                        param_opt_to_nec_pnads[option.parent], segment)
-                    assert ground_op is not None
+                    best_score_pnad, var_to_obj = \
+                        self._find_best_matching_pnad_and_sub(
+                        segment, objects,
+                        param_opt_to_nec_pnads[option.parent])
+                    assert var_to_obj is not None
                     assert best_score_pnad == pnad
-                    obj_to_var = dict(
-                        zip(ground_op.objects, pnad.op.parameters))
+                    obj_to_var = {v: k for k, v in var_to_obj.items()}
+                    assert len(var_to_obj) == len(obj_to_var)
+                    ground_op = pnad.op.ground(
+                        tuple(var_to_obj[var] for var in pnad.op.parameters))
 
                 # Every atom in the necessary_image that wasn't in the
```

Comment:
would be good to just verify by eye that every code path leading up to this point is correctly setting `obj_to_var` and that it's always consistent with `var_to_obj`

Reply:
i checked and it looks good!